runtime/lua/vim/diagnostic.lua

```diff
@@ -834,8 +848,82 @@ end
 ---                    diagnostics at the new cursor position (as if "cursor" were passed to
 ---                    the "scope" option).
 ---         - win_id: (number, default 0) Window ID
+---@deprecated
 function M.goto_next(opts)
-  return diagnostic_move_pos(opts, M.get_next_pos(opts))
+  vim.deprecate('vim.diagnostic.goto_next()', 'vim.diagnostic.goto()', '0.12')
+  local next = M.get_next(opts)
+  if next then
+    return diagnostic_move_pos(opts, { next.lnum, next.col })
+  end
+end
+
+--- @class vim.diagnostic.GotoOptions
+--- @field diagnostic Diagnostic?
+--- @field count number?
+--- @field namespace number?
+--- @field cursor_position number[]?
+--- @field wrap boolean?
+--- @field severity DiagnosticSeverity?
+--- @field float boolean|table|nil
+--- @field win_id number?
+
+--- Move to a diagnostic.
+---
+--- @param opts vim.diagnostic.GotoOptions A table with the following optional keys:
+---               - diagnostic: (table) The diagnostic item to move to |diagnostic-structure|.
+---                                     Mutually exclusive with "count", "namespace", and "severity".
+---               - count: (number) The number of diagnostics to move by. A positive integer moves
+---                                 forward by {count} diagnostics, while a negative integer moves
+---                                 backward by {count} diagnostics. Mutually exclusive with
+---                                 "diagnostic".
+---               - namespace: (number) Only consider diagnostics from the given namespace. Only used
+---                                     when "count" is non-nil.
+---               - cursor_position: (table) Cursor position as a (row, col) tuple. Used to find the
+---                                          nearest diagnostic when "count" is used. Only used when
+---                                          "count" is non-nil.
+---               - wrap: (boolean) Whether to wrap around the file or not. Similar to 'wrapscan'.
+---                                 Default true.
+---               - severity: Only consider diagnostics that match the given |diagnostic-severity|.
+---                           Only used when "count" is non-nil.
+---               - float: (boolean or table) If "true", call |vim.diagnostic.open_float()| after
+---                                           moving. If a table, also pass the table as the {opts}
+---                                           parameter to |vim.diagnostic.open_float()|. Unless
+---                                           overridden, the float will show diagnostics at the new
+---                                           cursor position (as if "cursor" were used as the
+---                                           "scope" option).
+---               - winid: (number) Window ID. Default 0 (current window).
```

Comment:
I think this should be `win_id`. See L496 (L501) in `diagnostic_move_pos`, this is already using `opts.win_id` so `winid` would be a breaking change.